# Coffee Quality Database from CQI (2nd Dataset)

## Imports/Pathways

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# File Pathway
path = r"C:\Users\Chase\anaconda_projects\Exercise_6_Coffee\A6_Coffee"

In [3]:
# Import 2nd dataset
df_cqi = pd.read_csv(
    r'C:\Users\Chase\anaconda_projects\Exercise_6_Coffee\A6_Coffee\02_Data\Original_Data\Coffee_Quality_database_from_CQI\Coffee_Quality_database_from_CQI_mc.csv')

## Data Inspection¶
### Table of Contents
1. Overview
2. Missing Values
3. Duplicate Check
4. Text Field Consistency Audit
5. Column Status: Keep, Drop, or Review

### 1. Overview

In [4]:
# General check of first 10 rows
df_cqi.head()

,Unnamed_Index',Species,Owner,Country.of.Origin,Farm.Name,Lot.Number,Mill,ICO.Number,Company,ICO.Number.1,...,Color,Category.Two.Defects,Expiration,Certification.Body,Certification.Address,Certification.Contact,unit_of_measurement,altitude_low_meters,altitude_high_meters,altitude_mean_meters
0,0,Arabica,metad plc,Ethiopia,metad plc,NaN,metad plc,2014/2015,metad agricultural developmet plc,1950-2200,...,Green,0,"April 3rd, 2016",METAD Agricultural Development plc,309fcf77415a3661ae83e027f7e5f05dad786e44,19fef5a731de2db57d16da10287413f5f99bc2dd,m,1950.0,2200.0,2075.0
1,1,Arabica,metad plc,Ethiopia,metad plc,NaN,metad plc,2014/2015,metad agricultural developmet plc,1950-2200,...,Green,1,"April 3rd, 2016",METAD Agricultural Development plc,309fcf77415a3661ae83e027f7e5f05dad786e44,19fef5a731de2db57d16da10287413f5f99bc2dd,m,1950.0,2200.0,2075.0
2,2,Arabica,grounds for health admin,Guatemala,"san marcos barrancas ""san cristobal cuch",NaN,NaN,NaN,NaN,1600 - 1800 m,...,NaN,0,"May 31st, 2011",Specialty Coffee Association,36d0d00a3724338ba7937c52a378d085f2172daa,0878a7d4b9d35ddbf0fe2ce69a2062cceb45a660,m,1600.0,1800.0,1700.0
3,3,Arabica,yidnekachew dabessa,Ethiopia,yidnekachew dabessa coffee plantation,NaN,wolensu,NaN,yidnekachew debessa coffee plantation,1800-2200,...,Green,2,"March 25th, 2016",METAD Agricultural Development plc,309fcf77415a3661ae83e027f7e5f05dad786e44,19fef5a731de2db57d16da10287413f5f99bc2dd,m,1800.0,2200.0,2000.0
4,4,Arabica,metad plc,Ethiopia,metad plc,NaN,metad plc,2014/2015,metad agricultural developmet plc,1950-2200,...,Green,2,"April 3rd, 2016",METAD Agricultural Development plc,309fcf77415a3661ae83e027f7e5f05dad786e44,19fef5a731de2db57d16da10287413f5f99bc2dd,m,1950.0,2200.0,2075.0


In [5]:
# Cehck the dataset shape
df_cqi.shape

(1339, 44)

In [7]:
# List all column headers
df_cqi.columns.tolist()

["Unnamed_Index'",
 'Species',
 'Owner',
 'Country.of.Origin',
 'Farm.Name',
 'Lot.Number',
 'Mill',
 'ICO.Number',
 'Company',
 'ICO.Number.1',
 'Region',
 'Producer',
 'Number.of.Bags',
 'Bag.Weight',
 'In.Country.Partner',
 'Harvest.Year',
 'Grading.Date',
 'Owner.1',
 'Variety',
 'Processing.Method',
 'Aroma',
 'Flavor',
 'Aftertaste',
 'Acidity',
 'Body',
 'Balance',
 'Uniformity',
 'Clean.Cup',
 'Sweetness',
 'Cupper.Points',
 'Total.Cup.Points',
 'Moisture',
 'Category.One.Defects',
 'Quakers',
 'Color',
 'Category.Two.Defects',
 'Expiration',
 'Certification.Body',
 'Certification.Address',
 'Certification.Contact',
 'unit_of_measurement',
 'altitude_low_meters',
 'altitude_high_meters',
 'altitude_mean_meters']

Headers will need work. Proper headers and standardize the text

In [8]:
# Check data types
df_cqi.dtypes

Unnamed_Index'             int64
Species                   object
Owner                     object
Country.of.Origin         object
Farm.Name                 object
Lot.Number                object
Mill                      object
ICO.Number                object
Company                   object
ICO.Number.1              object
Region                    object
Producer                  object
Number.of.Bags             int64
Bag.Weight                object
In.Country.Partner        object
Harvest.Year              object
Grading.Date              object
Owner.1                   object
Variety                   object
Processing.Method         object
Aroma                    float64
Flavor                   float64
Aftertaste               float64
Acidity                  float64
Body                     float64
Balance                  float64
Uniformity               float64
Clean.Cup                float64
Sweetness                float64
Cupper.Points            float64
Total.Cup.

### 2. Missing Values

In [9]:
# Create a summary table of missing values for CQI dataset, excluding 0s
missing_count = df_cqi.isnull().sum()
missing_percent = (df_cqi.isnull().mean() * 100).round(2)

missing_summary = pd.DataFrame({
    'Missing Count': missing_count,
    'Missing %': missing_percent
})

# Filter out rows with 0 missing values
missing_summary = missing_summary[missing_summary['Missing Count'] > 0]

# Sorted by highest missing %
missing_summary = missing_summary.sort_values(by='Missing %', ascending=False)

# Display the summary
missing_summary

,Missing Count,Missing %
Lot.Number,1063,79.39
Farm.Name,359,26.81
Mill,318,23.75
Color,270,20.16
Producer,232,17.33
altitude_high_meters,230,17.18
altitude_low_meters,230,17.18
altitude_mean_meters,230,17.18
ICO.Number.1,226,16.88
Variety,226,16.88


Looking at this I will be dropping the Lot.Number column, too high % missing to be useful. The rest will be inspected for what can be done.

### 3. Duplicate Check

In [10]:
# Check for duplicate rows
df_cqi.duplicated().sum()

np.int64(0)

No duplicated rows found.

### 4. Text Field Consistency Audit

#### Headers

In [11]:
# Headers are inconsistent 
df_cqi.columns.tolist()

["Unnamed_Index'",
 'Species',
 'Owner',
 'Country.of.Origin',
 'Farm.Name',
 'Lot.Number',
 'Mill',
 'ICO.Number',
 'Company',
 'ICO.Number.1',
 'Region',
 'Producer',
 'Number.of.Bags',
 'Bag.Weight',
 'In.Country.Partner',
 'Harvest.Year',
 'Grading.Date',
 'Owner.1',
 'Variety',
 'Processing.Method',
 'Aroma',
 'Flavor',
 'Aftertaste',
 'Acidity',
 'Body',
 'Balance',
 'Uniformity',
 'Clean.Cup',
 'Sweetness',
 'Cupper.Points',
 'Total.Cup.Points',
 'Moisture',
 'Category.One.Defects',
 'Quakers',
 'Color',
 'Category.Two.Defects',
 'Expiration',
 'Certification.Body',
 'Certification.Address',
 'Certification.Contact',
 'unit_of_measurement',
 'altitude_low_meters',
 'altitude_high_meters',
 'altitude_mean_meters']

Headers will need to be lowercase and have an underscore between the words.Spelling check too.

#### Casing and Spacing Consistency

In [12]:
# Inspect all object-type columns for unique values
for col in df_cqi.select_dtypes(include='object').columns:
    print(f"\n{col} unique values:")
    print(df_cqi[col].dropna().unique())


Species unique values:
['Arabica' 'Robusta']

Owner unique values:
['metad plc' 'grounds for health admin' 'yidnekachew dabessa' 'ji-ae ahn'
 'hugo valdivia' 'ethiopia commodity exchange' 'diamond enterprise plc'
 'mohammed lalo' 'cqi q coffee sample representative'
 'yunnan coffee exchange' 'essencecoffee' 'the coffee source inc.'
 'roberto licona franco' 'nucoffee' 'kabum trading company'
 'bismarck castro' 'lin, che-hao krude 林哲豪' 'nora zeas'
 'specialty coffee-korea' 'francisco a mena' 'hider abamecha'
 'daniel magu' 'kona pacific farmers cooperative' 'itdp international'
 'jacques pereira carneiro' 'jungle estate' 'great lakes coffee uganda'
 'lusso lab' 'afca' 'juan luis alvarado romero' 'kawacom uganda ltd'
 'exportadora de cafe condor s.a' 'gonzalo hernandez'
 'ibrahim hussien speciality coffee producer &export'
 'seid damtew coffee planataion' 'dane loraas' 'colbran coffeelands, ltd.'
 'atlantic specialty coffee' 'assefa belay coffee producer'
 'kyagalanyi ltd' 'rashid moledi

Several text fields contain inconsistent casing, spacing, and potential spelling variants. Some entries include non-English characters or long compound names. These will need normalization, typo correction, and possible grouping or exclusion during the cleaning phase.

#### Encoding and Character Issues

In [13]:
df_cqi['Certification.Contact'].sample(10)

88     19fef5a731de2db57d16da10287413f5f99bc2dd
895    7f521ca403540f81ec99daec7da19c2788393880
236    0878a7d4b9d35ddbf0fe2ce69a2062cceb45a660
29     0878a7d4b9d35ddbf0fe2ce69a2062cceb45a660
168    724f04ad10ed31dbb9d260f0dfd221ba48be8a95
17     61bbaf6a9f341e5782b8e7bd3ebf76aac89fe24b
702    aa2ff513ffb9c844462a1fb07c599bce7f3bb53d
803    5eb2b7129d9714c43825e44dc3bca9423de209e9
196    f6d87a6c04653c569d4911a66f89d5e30ce83b93
553    0eb4ee5b3f47b20b049548a2fd1e7d4a2b70d0a7
Name: Certification.Contact, dtype: object

This column, certification.contact will be dropped.

#### Unique Value Scan

In [14]:
# Count unique values in all object-type columns
df_cqi.select_dtypes(include='object').nunique().sort_values(ascending=False)

ICO.Number               845
Producer                 692
Farm.Name                571
Grading.Date             567
Expiration               566
Mill                     459
ICO.Number.1             396
Region                   356
Owner.1                  319
Owner                    315
Company                  281
Lot.Number               227
Bag.Weight                56
Harvest.Year              46
Country.of.Origin         36
Certification.Address     32
Variety                   29
Certification.Contact     29
In.Country.Partner        27
Certification.Body        26
Processing.Method          5
Color                      3
Species                    2
unit_of_measurement        2
dtype: int64

Some fields have too many unique values for grouping; others show clean, low-cardinality categories ideal for analysis.

#### Input Normalization

In [37]:
# Checking for inconsistencies
# 1. origin_country
print("origin_country:")
print(df_cqi['origin_country'].value_counts(dropna=False).sort_index(), end="\n\n")

origin_country:
origin_country
brazil                          132
burundi                           2
china                            16
colombia                        183
costa rica                       51
cote d?ivoire                     1
ecuador                           3
el salvador                      21
ethiopia                         44
guatemala                       181
haiti                             6
honduras                         53
india                            14
indonesia                        20
japan                             1
kenya                            25
laos                              3
malawi                           11
mauritius                         1
mexico                          236
myanmar                           8
nicaragua                        26
panama                            4
papua new guinea                  1
peru                             10
philippines                       5
rwanda                           

origin_country note:
Mostly clean and lowercase.
Fix encoding in "cote d?ivoire" to "côte d'ivoire".
Simplify "tanzania, united republic of" to "tanzania".
Collapse "united states (hawaii)" to "united states".
Treat "puerto rico" separately from "united states" if needed.

In [34]:
# Checking for inconsistencies
# 2. region
print("region:")
print(df_cqi['region'].value_counts(dropna=False).sort_index(), end="\n\n")

region:
region
52 narino (exact location: mattituy; municipal region: florida code 381     1
acatenango                                                                  3
aceh                                                                        1
aceh gayo                                                                   1
aceh tengah                                                                 1
                                                                           ..
嘉義阿里山                                                                       2
國姓鄉 guoshing township                                                       2
苗栗三灣                                                                        1
苗栗泰安                                                                        1
NaN                                                                        59
Name: count, Length: 357, dtype: int64



region note:
Highly fragmented—357 distinct values with many rare entries.
Contains detailed location strings, multilingual entries, and inconsistent formatting.
No normalization recommended without a reference table or geographic hierarchy.

In [35]:
# Checking for inconsistencies
# 3. variety
print("variety:")
print(df_cqi['variety'].value_counts(dropna=False).sort_index(), end="\n\n")

variety:
variety
arusha                     6
blue mountain              2
bourbon                  226
catimor                   20
catuai                    74
caturra                  256
ethiopian heirlooms        1
ethiopian yirgacheffe      2
gesha                     12
hawaiian kona             44
java                       2
mandheling                 3
marigojipe                 1
moka peaberry              1
mundo novo                33
other                    110
pacamara                   8
pacas                     13
pache comun                1
peaberry                   5
ruiru 11                   2
sl14                      17
sl28                      15
sl34                       8
sulawesi                   1
sumatra                    3
sumatra lintong            1
typica                   211
yellow bourbon            35
NaN                      226
Name: count, dtype: int64



variety note:
Mostly clean and domain-consistent.
No clear typos or casing issues.
“other” is broad—consider flagging for manual review.
Rare entries like “marigojipe” and “sumatra lintong” are valid but may benefit from grouping if analysis requires simplification.
No normalization needed unless varietal grouping is part of your analysis plan.


In [36]:
# Checking for inconsistencies
# 4. species
print("species:")
print(df_cqi['species'].value_counts(dropna=False).sort_index())

species:
species
arabica    1311
robusta      28
Name: count, dtype: int64


species note:
Clean and consistent.
Only includes “arabica” and “robusta” with no typos or formatting issues.
No normalization needed.

### 5. Column Status: Keep, Drop, or Review

#### Keep

In [15]:
columns_to_keep = [
    'Species',
    'Country.of.Origin',
    'Harvest.Year',
    'Processing.Method',
    'Color',
    'unit_of_measurement',
    'In.Country.Partner',
    'Certification.Body',
    'Variety',
    'Number.of.Bags',
    'Bag.Weight',
    'Aroma',
    'Flavor',
    'Aftertaste',
    'Acidity',
    'Body',
    'Balance',
    'Uniformity',
    'Clean.Cup',
    'Sweetness',
    'Cupper.Points',
    'Total.Cup.Points',
    'altitude_mean_meters',
    'Region',
    'Moisture',
    ]

These are clean, interpretable, and aligned with my analysis goals. They support origin context, production methods, and filtering logic. Low cardinality, clear semantics, and strong potential for enrichment—these are staying in.


#### Drop

In [16]:
columns_to_drop = [
    'ICO.Number',
    'Certification.Contact',
    'Certification.Address',
    'Producer',
    'Farm.Name',
    'Mill',
    'Company',
    'Lot.Number',
    'Grading.Date',
    'Expiration',
    'Owner.1',
    'Category.One.Defects',
    'Category.Two.Defects'
]

These fields are being dropped due to high cardinality, unclear context, or lack of analytical value. Most are unique identifiers, fragmented text, or fields without interpretable meaning. They don’t support grouping, enrichment, or storytelling, so they’re out.


#### Review

In [17]:
columns_to_review = [
    'Owner',
    'Region',
    'Moisture',
    'Quakers',
    'altitude_low_meters',
    'altitude_high_meters'
]

These are on the fence. Some might be salvageable with normalization or clarification, while others need unit context or grading logic. I’m holding off on decisions here until I’ve looked into this more.


## Data Cleaning
### Table of Contents
1. Dropping Columns
2. Standardize Headers and Text Fields
3. Normalize Inputs (needed????)
4. Missing Values
5. Export

### 1. Dropping Columns

After reviewing all the columns, I have decided on the following columns to be dropped.


In [18]:
columns_to_drop = [
    "Unnamed_Index'",
    'Farm.Name',
    'Lot.Number',
    'Mill',
    'ICO.Number',
    'Company',
    'ICO.Number.1',
    'Producer',
    'Number.of.Bags',
    'Bag.Weight',
    'Grading.Date',
    'Owner.1',
    'Category.One.Defects',
    'Quakers',
    'Color',
    'Category.Two.Defects',
    'Expiration',
    'Certification.Body',
    'Certification.Address',
    'Certification.Contact',
    'altitude_low_meters',
    'altitude_high_meters'
]

df_cqi = df_cqi.drop(columns=columns_to_drop)

In [19]:
# Work Check
df_cqi.columns.tolist()

['Species',
 'Owner',
 'Country.of.Origin',
 'Region',
 'In.Country.Partner',
 'Harvest.Year',
 'Variety',
 'Processing.Method',
 'Aroma',
 'Flavor',
 'Aftertaste',
 'Acidity',
 'Body',
 'Balance',
 'Uniformity',
 'Clean.Cup',
 'Sweetness',
 'Cupper.Points',
 'Total.Cup.Points',
 'Moisture',
 'unit_of_measurement',
 'altitude_mean_meters']

### 2. Standardize Headers and Text Fields

#### Headers

In [22]:
# Standardize Header Format
df_cqi.columns = (
    df_cqi.columns
    .str.strip()
    .str.lower()
    .str.replace('\n', ' ')
    .str.replace('\r', '')
    .str.replace('.', '_')
)

In [23]:
# Work check
df_cqi.columns.tolist()

['species',
 'owner',
 'country_of_origin',
 'region',
 'in_country_partner',
 'harvest_year',
 'variety',
 'processing_method',
 'aroma',
 'flavor',
 'aftertaste',
 'acidity',
 'body',
 'balance',
 'uniformity',
 'clean_cup',
 'sweetness',
 'cupper_points',
 'total_cup_points',
 'moisture',
 'unit_of_measurement',
 'altitude_mean_meters']

In [24]:
# Rename headers to be consistant 
df_cqi.rename(columns={
    'country_of_origin': 'origin_country',
    'in_country_partner': 'local_partner'
}, inplace=True)

In [25]:
# Work check
df_cqi.columns.tolist()

['species',
 'owner',
 'origin_country',
 'region',
 'local_partner',
 'harvest_year',
 'variety',
 'processing_method',
 'aroma',
 'flavor',
 'aftertaste',
 'acidity',
 'body',
 'balance',
 'uniformity',
 'clean_cup',
 'sweetness',
 'cupper_points',
 'total_cup_points',
 'moisture',
 'unit_of_measurement',
 'altitude_mean_meters']

#### Standardize Text Field Values

In [27]:
# ID text columns
text_columns = df_cqi.select_dtypes(include='object').columns.tolist()

In [28]:
# Define Standardization Function
def standardize_text(text):
    if pd.isna(text):
        return text
    return (
        str(text)
        .strip()                     # Remove leading/trailing whitespace
        .lower()                     # Convert to lowercase
        .replace('\n', ' ')          # Remove line breaks
        .replace('\r', '')           # Remove carriage returns
    )

In [29]:
# Apply to text columns
df_cqi[text_columns] = df_cqi[text_columns].applymap(standardize_text)

C:\Users\Chase\AppData\Local\Temp\ipykernel_26032\3722561294.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df_cqi[text_columns] = df_cqi[text_columns].applymap(standardize_text)


In [30]:
# Work check
df_cqi[['species', 'origin_country', 'processing_method']].sample(5)

,species,origin_country,processing_method
1214,arabica,nicaragua,NaN
167,arabica,mexico,washed / wet
1008,arabica,uganda,natural / dry
565,arabica,guatemala,washed / wet
551,arabica,guatemala,washed / wet


### 3. Normalize Inputs

In [38]:
# Adressing 1. origin_country
# Normalize origin_country values
origin_country_map = {
    "cote d?ivoire": "côte d'ivoire",
    "tanzania, united republic of": "tanzania",
    "united states (hawaii)": "united states",
    "united states (puerto rico)": "puerto rico"
}

df_cqi['origin_country'] = df_cqi['origin_country'].replace(origin_country_map)

In [40]:
# Work check
print(df_cqi['origin_country'].value_counts(dropna=False).sort_index())

origin_country
brazil              132
burundi               2
china                16
colombia            183
costa rica           51
côte d'ivoire         1
ecuador               3
el salvador          21
ethiopia             44
guatemala           181
haiti                 6
honduras             53
india                14
indonesia            20
japan                 1
kenya                25
laos                  3
malawi               11
mauritius             1
mexico              236
myanmar               8
nicaragua            26
panama                4
papua new guinea      1
peru                 10
philippines           5
puerto rico           4
rwanda                1
taiwan               75
tanzania             40
thailand             32
uganda               36
united states        83
vietnam               8
zambia                1
NaN                   1
Name: count, dtype: int64


### 4. Missing Values

In [42]:
# Create a summary table of missing values for CQI dataset, excluding 0s
missing_count = df_cqi.isnull().sum()
missing_percent = (df_cqi.isnull().mean() * 100).round(2)

missing_summary = pd.DataFrame({
    'Missing Count': missing_count,
    'Missing %': missing_percent
})

# Filter out rows with 0 missing values
missing_summary = missing_summary[missing_summary['Missing Count'] > 0]

# Sorted by highest missing %
missing_summary = missing_summary.sort_values(by='Missing %', ascending=False)

# Display the summary
missing_summary

,Missing Count,Missing %
altitude_mean_meters,230,17.18
variety,226,16.88
processing_method,170,12.70
region,59,4.41
harvest_year,47,3.51
owner,7,0.52
origin_country,1,0.07


Several columns contain missing values that cannot be reliably imputed. To preserve transparency and analytical integrity:
- Numeric columns (altitude_mean_meters, harvest_year) will remain as NaN to avoid introducing bias or distortion.
- String columns (variety, processing_method, owner) will be filled with "Unknown" to maintain categorical structure while signaling absence.
- This approach ensures the dataset remains unskewed, auditable, and ready for downstream analysis.
- I will investigate region and origin_country further to determine whether missing values can be inferred based on known patterns or relationships.


In [43]:
# Display row where origin_country is missing
missing_country_row = df_cqi[df_cqi['origin_country'].isna()]
print(missing_country_row)

      species               owner origin_country region local_partner  \
1197  arabica  racafe & cia s.c.a            NaN    NaN      almacafé   

     harvest_year variety processing_method  aroma  flavor  ...  body  \
1197          NaN     NaN               NaN   6.75    6.75  ...  7.58   

      balance  uniformity  clean_cup  sweetness  cupper_points  \
1197      7.5        10.0       10.0       10.0           7.25   

      total_cup_points  moisture  unit_of_measurement  altitude_mean_meters  
1197             79.08       0.1                    m                   NaN  

[1 rows x 22 columns]


In [46]:
# Filter rows with same owner: racafe & cia s.c.a to see if one has a origin_country entry
racafe_rows = df_cqi[
    (df_cqi['owner'].str.lower() == 'racafe & cia s.c.a') &
    (df_cqi['origin_country'].notna())
]

# Display relevant columns
print(racafe_rows[['origin_country', 'region']])

     origin_country     region
97         colombia      huila
106        colombia      huila
107        colombia      huila
110        colombia      huila
145        colombia  santander
146        colombia      huila
151        colombia      huila
174        colombia      huila
183        colombia   la plata
197        colombia      huila
198        colombia      huila
211        colombia      huila
216        colombia      huila
235        colombia      huila
249        colombia      huila
268        colombia      huila
275        colombia      huila
299        colombia      huila
304        colombia      huila
325        colombia      huila
337        colombia      huila
339        colombia      huila
347        colombia      huila
356        colombia      huila
371        colombia      huila
397        colombia    guayata
398        colombia      huila
436        colombia   la plata
455        colombia      huila
467        colombia      huila
479        colombia      huila
492     

In [47]:
# From the above list it shows the company is out of Colombia. I will enter this data in
# Fill origin_country for the Racafé row
df_cqi.loc[df_cqi['owner'].str.lower() == 'racafe & cia s.c.a', 'origin_country'] = 'colombia'

In [49]:
# Work check
# Count missing values in origin_country
missing_origin_country = df_cqi['origin_country'].isna().sum()
print(f"Missing origin_country values: {missing_origin_country}")

Missing origin_country values: 0


In [50]:
# Display origin_country and region side by side to investigate regions
country_region_view = df_cqi[['origin_country', 'region']].sort_values(by=['origin_country', 'region'])
print(country_region_view)

     origin_country                     region
987          brazil  alta paulista (sao paulo)
734          brazil      brazil matas de minas
421          brazil     campos altos - cerrado
637          brazil     campos altos - cerrado
681          brazil     campos altos - cerrado
...             ...                        ...
860         vietnam                    vietnam
502         vietnam            vietnam cau dat
564         vietnam              vietnam tutra
1338        vietnam                        NaN
823          zambia              mubuyu estate

[1339 rows x 2 columns]


In [45]:
# Filter rows with missing region but known origin_country
missing_region_candidates = df_cqi[df_cqi['region'].isna() & df_cqi['origin_country'].notna()]
print(missing_region_candidates[['origin_country', 'region']])

     origin_country region
2         guatemala    NaN
5            brazil    NaN
6              peru    NaN
30        nicaragua    NaN
31         tanzania    NaN
100           kenya    NaN
114           kenya    NaN
134      costa rica    NaN
152           kenya    NaN
223       guatemala    NaN
240      costa rica    NaN
257        thailand    NaN
293        thailand    NaN
334   united states    NaN
352          taiwan    NaN
373   united states    NaN
379          taiwan    NaN
394        thailand    NaN
407            peru    NaN
415   united states    NaN
443            peru    NaN
457        colombia    NaN
495       guatemala    NaN
521        colombia    NaN
526        thailand    NaN
528   united states    NaN
529   united states    NaN
532        thailand    NaN
548       guatemala    NaN
558        thailand    NaN
623     el salvador    NaN
626          mexico    NaN
632        colombia    NaN
645       guatemala    NaN
690        honduras    NaN
735          taiwan    NaN
7

After reviewing the country–region pairs, I’ve determined that I cannot safely impute region based on origin_country. Many countries have multiple distinct regions or highly specific entries, and forcing a single match would risk skewing the analysis. To preserve transparency and analytical integrity, missing region values will be filled with - unknown.


In [53]:
# Define string columns to fill
string_cols = ['variety', 'processing_method', 'owner', 'region']

# Fill missing string values with "Unknown"
df_cqi[string_cols] = df_cqi[string_cols].fillna('unknown')

In [54]:
# Work check
# Show missing value counts for string columns
string_cols = ['variety', 'processing_method', 'owner', 'region']
print(df_cqi[string_cols].isna().sum())

variety              0
processing_method    0
owner                0
region               0
dtype: int64


Unknown imputation was successful 

In [55]:
# Show missing value counts for numeric columns
numeric_cols = ['altitude_mean_meters', 'harvest_year']
print(df_cqi[numeric_cols].isna().sum())

altitude_mean_meters    230
harvest_year             47
dtype: int64


This is fine since I'm leaving as NaN

### 5. Export

In [57]:
# Export cleaned CQI dataset
df_cqi.to_csv(r"C:\Users\Chase\anaconda_projects\Exercise_6_Coffee\A6_Coffee\02_Data\Prepared_Data\Coffee_Quality_database_from_CQI\coffee_quality_cqi_clean.csv", index=False)